addons/base_automation/models/base_automation.py

```diff
@@ -1006,22 +1032,54 @@ def get_record_dt(record):
+                relative_now = now + relative_value
+                try:
+                    relative_last_run = last_run + relative_value
+                except OverflowError:
+                    relative_last_run = datetime.datetime.fromtimestamp(0, tz=None)
+                time_domain = Domain(dt_field.name, '>=', relative_last_run) & Domain(dt_field.name, '<', relative_now)
+                if is_date_automation_last:
+                    time_domain |= Domain(dt_field.name, '=', False) & Domain('create_date', '>=', relative_last_run) & Domain('create_date', '<', relative_now)
+
+                if (dt_field.store or dt_field.search):
+                    records = Model.search(time_domain & domain)
                 else:
-                    is_process_to_run = (
-                        last_run <=
-                        fields.Datetime.from_string(record_dt) + DATE_RANGE_FUNCTION[automation.trg_date_range_type](automation.trg_date_range)
-                        < now
-                    )
-                if is_process_to_run:
-                    try:
+                    records = Model.search(domain).filtered_domain(time_domain)
+
+                if len(records) > batch_target_size > 0:
+                    # we have many records, limit now to a date close to the batch_target_size
+                    original_records = records
+                    records = records.sorted(get_record_dt)
+                    now = get_record_dt(records[batch_target_size - 1])
+                    relative_now = now + relative_value
+                    records = records.filtered(lambda rec: get_record_dt(rec) <= relative_now)
+                    if len(original_records) > len(records):
+                        partial_runs += 1
+                        _logger.debug("Process partial set of records in time-based automation")
+
+            # run the automation on the records
+            for record in records:
+                try:
+                    with self.env.cr.savepoint():
                         automation._process(record)
-                    except Exception:
-                        _logger.error(traceback.format_exc())
+                except Exception:
+                    _logger.error(traceback.format_exc())
 
-            automation.write({'last_run': now.strftime(DEFAULT_SERVER_DATETIME_FORMAT)})
+            automation.write({'last_run': now})
+            self.env.flush_all()
             _logger.info("Time-based automation rule `%s` done.", automation.name)
 
             if automatic:
                 # auto-commit for batch processing
-                self._cr.commit()
+                # because of `last_run`, we must process all records in the batch
+                self.env['ir.cron']._notify_progress(
+                    done=automation_number - partial_runs,
+                    remaining=(len(automations) - automation_number + partial_runs),
+                )
+                self.env.cr.commit()
```

Comment:
Same comment: we should notify at each step of the loop instead of using a savepoint.

Reply:
Because of the algorithm assumes that all records between "last run" and "now" are processed atomically per automation, we cannot commit at each loop.
If we want to process each record in a separate transaction, we would need to manage a queue of jobs to execute. (not in scope in this PR)

Reply:
OK, got it.
I think this one is tricky: the real issue here is that a timeout is often the consequence of a (single) poorly written automation. The question becomes: do we really want to make a lot of effort to process as much as we can, knowing that at some point we will probably never be able to process all records anyway.
I guess that ideally, we should be able to deactivate an automation individually, but that may be too much effort to compensate some bad code.

Reply:
It would require some more advanced mechanism to solve all the issues. A queue would be an easier solution in the long term.
In the meantime, this refactoring introduces some performance improvements because it avoids fetching all records from the database (when possible).